maro/simulator/scenarios/supply_chain/units/storage.py

```diff
@@ -1,51 +1,219 @@
+            product_id (int): Product to check.
 
         Returns:
-            Dict[int, int]: Dictionary of product id and quantity success added.
+            int: Available quantity of product.
         """
-        if all_or_nothing and self.remaining_space < sum(product_quantities.values()):
-            return {}
+        return self._product_level[product_id]
 
-        unloaded_quantities: Dict[int, int] = {}
+    def get_product_max_remaining_space(self, product_id: int) -> int:
+        # ! Currently the upper bound is only used in Manufacture Unit.
+        remaining_space = self._remaining_space_dict[self._product2storage[product_id]]
 
-        for product_id, quantity in product_quantities.items():
-            unload_quantity = min(self.remaining_space, quantity)
+        upper_bound = self._storage_sku_upper_bound[self._product2storage[product_id]][product_id]
+        remaining_quota = max(upper_bound - self._product_level[product_id], 0)
+
+        return min(remaining_space, remaining_quota)
 
-            self._product_level[product_id] += unload_quantity
-            unloaded_quantities[product_id] = unload_quantity
+    def _add_product(self, product_id: int, quantity: int) -> None:
+        assert self._remaining_space_dict[self._product2storage[product_id]] >= quantity
+        self._product_level[product_id] += quantity
+        self._product_level_changed[product_id] = True
+        self._remaining_space_dict[self._product2storage[product_id]] -= quantity
 
-            self._changed_product_cache[product_id] = True
+    def _take_product(self, product_id: int, quantity: int) -> None:
+        assert self._product_level[product_id] >= quantity
+        self._product_level[product_id] -= quantity
+        self._product_level_changed[product_id] = True
+        self._remaining_space_dict[self._product2storage[product_id]] += quantity
+
+    def try_add_products(
+        self,
+        product_quantities: Dict[int, int],
+        add_strategy: AddStrategy=AddStrategy.IgnoreUpperBoundAllOrNothing
+    ) -> Dict[int, int]:
+        """Try to add products into storage.
 
-            self.remaining_space -= unload_quantity
+        Args:
+            product_quantities (Dict[int, int]): Dictionary of product id and quantity need to add to storage.
+            add_strategy (AddStrategy): The strategy to add products into the storage, Defaults to AddAllOrNothing.
 
-        return unloaded_quantities
+        Returns:
+            Dict[int, int]: Dictionary of product id and quantity success added.
+        """
+        added_quantities: Dict[int, int] = {}
+
+        if add_strategy in [AddStrategy.IgnoreUpperBoundAllOrNothing, AddStrategy.IgnoreUpperBoundProportional]:
+            space_requirements: Dict[int, int] = {}
+            for product_id, quantity in product_quantities.items():
+                storage_id = self._product2storage[product_id]
+                if storage_id not in space_requirements:
```

Comment:
Better to use a `defaultdict`

Reply:
update `self._storage_sku_upper_bound` accordingly